src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -183,56 +218,104 @@ protected override void Dispose(bool disposing)
             }
 
             // Create RequestCompletionSource object and save current values of handler settings.
-            RequestCompletionSource state = new RequestCompletionSource();
-            state.CancellationToken = cancellationToken;
-            state.RequestMessage = request;
-            state.Handler = this;
+            RequestCompletionSource state = new RequestCompletionSource
+            {
+                CancellationToken = cancellationToken,
+                RequestMessage = request,
+            };
 
-            Task.Factory.StartNew(
-                s =>
-                {
-                    var rcs = (RequestCompletionSource)s;
-                    rcs.Handler.StartRequest(rcs);
-                },
-                state,
-                CancellationToken.None,
-                TaskCreationOptions.DenyChildAttach,
-                TaskScheduler.Default);
+            BeginRequest(state);
 
             return state.Task;
         }
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void BeginRequest(RequestCompletionSource state)
         {
-            HttpResponseMessage responseMessage = null;
-            SafeCurlHandle requestHandle = null;
+            SafeCurlHandle requestHandle = new SafeCurlHandle();
+            GCHandle stateHandle = new GCHandle();
+            bool needCleanup = false;
 
             if (state.CancellationToken.IsCancellationRequested)
             {
-                state.TrySetCanceled();
+                state.TrySetCanceled(state.CancellationToken);
                 return;
             }
 
-            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
-
             try
             {
-                requestHandle = CreateRequestHandle(state);
+                // Prepare context objects
+                stateHandle = GCHandle.Alloc(state);
+                state.ResponseMessage = new CurlResponseMessage(state.RequestMessage);
+                requestHandle = CreateRequestHandle(state, stateHandle);
+                state.RequestHandle = requestHandle;
+                needCleanup = true;
+
+                if (state.CancellationToken.IsCancellationRequested)
+                {
+                    state.TrySetCanceled(state.CancellationToken);
+                    return;
```

Comment:
Sorry.. I removed the earlier finally block without checking carefully. Fixed this now